### PR TITLE
Feat(front): switch storage

### LINKED ISSUE #16: Ability to dynamically switch the storage
Actually (should be implemented in #12) the front end force local storage or SQLite for data storage. Make it customizable with a switch in a react context and store it in a key in each storages when updating. 

On boot, the frontend must check the data backend storage.

### PR DESCRIPTION
**Pull request by tomyo002**

### CODE REVIEW

File: frontend/src/context/StorageProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+export const storageContext = createContext<storageContextProps>({
+  storage: 'localStorage',
+  setStorage: storage => console.log(storage),
+});
+
+export const StorageProvider: React.FC<React.PropsWithChildren> = ({
+  children,
+}) => {
+  const [storage, setStorage] = useState<AllowedStorage>('localStorage');
```
Comment (darkweak) [inline frontend/src/context/StorageProvider.tsx]:
```suggestion
  const [storage, setStorage] = useState<IStorage>(new InstanceData());
  const updateStorage(name: string) {
    setStorage(new InstanceData(name));
  }
```

File: frontend/src/service/class/Storage.tsx [hunk 2]
```diff
@@ -35,4 +36,32 @@ export class SqliteStorage extends AbstractStorage {
...
+  }
+
+  get(storage: AllowedStorage) {
+    switch (storage) {
+      case 'sqliteStorage':
+        return this.sqliteStorage.get();
+      case 'localStorage':
+      default:
+        return this.localStorage.get();
+    }
```
Comment (darkweak) [inline frontend/src/service/class/Storage.tsx]:
```suggestion
    let current: IStorage;
    switch (storage) {
      case 'sqliteStorage':
        current = this.sqliteStorage;
        break;
      default:
        current = this.localStorage;
    }
    
    return current.get();
```

**Comment by tomyo002** [comment]
Avant
Capture d’écran du 2024-05-24 10-37-45
Apres
Capture d’écran du 2024-05-24 10-38-10

File: frontend/src/context/StorageProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+});
+
+export const StorageProvider: React.FC<React.PropsWithChildren> = ({
+  children,
+}) => {
+  const [storage, setStorage] = useState<IStorage>(
+    new InstanceData('localStorage'),
+  );
+  const [currentType, setCurrentType] =
+    useState<AllowedStorage>('localStorage');
```
Comment (darkweak) [inline frontend/src/context/StorageProvider.tsx]: Useless.

File: frontend/src/components/page/HomePage.tsx [hunk 4]
```diff
@@ -1,34 +1,64 @@
...
+  useAllInstances,
+  useSelectStorage,
+  useSetInstances,
+} from 'src/context';
 import { path } from 'src/utils';
 
 import { ButtonOutline, H1 } from '../atomic';
 import { Layout } from '../layout';
 
+const allowStorage = (value: string): value is AllowedStorage => {
```
Comment (darkweak) [inline frontend/src/components/page/HomePage.tsx]: That's an `is`-er function.

File: frontend/src/context/StorageProvider.tsx [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      value={{ storage, currentType, setStorage: updateStorage }}
+    >
+      {children}
+    </storageContext.Provider>
+  );
+};
+
+export const useStorage = () => {
+  const { storage } = useContext(storageContext);
+  return storage;
```
Comment (darkweak) [inline frontend/src/context/StorageProvider.tsx:35]: Missing new line.

File: frontend/src/context/StorageProvider.tsx [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+};
+
+export const useSelectStorage = () => {
+  const { currentType, setStorage } = useContext(storageContext);
+  return { currentType, setStorage };
```
Comment (darkweak) [inline frontend/src/context/StorageProvider.tsx]: Missing new line.

File: frontend/src/service/class/Storage.tsx [hunk 7]
```diff
@@ -35,4 +38,26 @@ export class SqliteStorage extends AbstractStorage {
...
+  constructor(storage: AllowedStorage) {
+    switch (storage) {
+      case 'sqliteStorage':
+        this.currentStorage = new SqliteStorage();
+        break;
+      default:
+        this.currentStorage = new LocalStorage();
+    }
+  }
+  async get() {
```
Comment (darkweak) [inline frontend/src/service/class/Storage.tsx:61]: Missing new line.

File: frontend/src/service/class/Storage.tsx [hunk 8]
```diff
@@ -35,4 +38,26 @@ export class SqliteStorage extends AbstractStorage {
...
+    return this.currentStorage.get();
+  }
+  async set(instances: readonly InstanceType[]) {
```
Comment (darkweak) [inline frontend/src/service/class/Storage.tsx]: Missing new line.

Comment (darkweak) [reply frontend/src/service/class/Storage.tsx]: Why not using the ReadonlyArray type?

File: frontend/src/service/class/Storage.tsx [hunk 9]
```diff
@@ -35,4 +38,26 @@ export class SqliteStorage extends AbstractStorage {
...
+    this.currentStorage.set(instances);
+  }
+  async delete() {
```
Comment (darkweak) [inline frontend/src/service/class/Storage.tsx:69]: Missing new line.

**Review by darkweak — CHANGES_REQUESTED** [review]

File: frontend/src/components/page/HomePage.tsx [hunk 10]
```diff
@@ -1,34 +1,64 @@
...
-        </>
-      )}
+        </Link>
+        {!!useAllInstances().length && (
+          <>
+            <ButtonOutline
+              className="btn-error w-full"
+              onClick={() => {
+                setInstances([]);
+              }}
```
Comment (darkweak) [inline frontend/src/components/page/HomePage.tsx]: Don't need extra curly brackets.

File: frontend/src/context/StorageProvider.tsx [hunk 11]
```diff
@@ -0,0 +1,44 @@
+import React, { createContext, useContext, useState } from 'react';
+import { InstanceData } from 'src/service/class';
+import { IStorage } from 'src/service/interface';
+
+export type AllowedStorage = 'localStorage' | 'sqliteStorage';
```
Comment (darkweak) [inline frontend/src/context/StorageProvider.tsx]: Store these as const to avoid typos.

**Review by darkweak — CHANGES_REQUESTED** [review]

File: frontend/src/context/StorageProvider.tsx [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+import { IStorage } from 'src/service/interface';
+import { AllowedStorage } from 'src/utils';
+
+type storageContextProps = {
+  storage: IStorage;
+  setStorage: (storage: AllowedStorage) => void;
+};
+
+const storageContext = createContext<storageContextProps>({
+  storage: new InstanceData('localStorage'),
```
Comment (darkweak) [inline frontend/src/context/StorageProvider.tsx]: Use the constant.

File: frontend/src/context/StorageProvider.tsx [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+  setStorage: storage => console.log(storage),
+});
+
+export const StorageProvider: React.FC<React.PropsWithChildren> = ({
+  children,
+}) => {
+  const [storage, setStorage] = useState<IStorage>(
+    new InstanceData('localStorage'),
```
Comment (darkweak) [inline frontend/src/context/StorageProvider.tsx]: Use the constant.

File: frontend/src/service/instance.tsx [hunk 14]
```diff
@@ -1,13 +1,16 @@
...
+  storage.set(instances);
 }
 
-export function getAllInstances(): Promise<ReadonlyArray<InstanceType>> {
-  return instanceStorage.get();
+export function getAllInstances(
+  storage: IStorage,
+): Promise<ReadonlyArray<InstanceType>> {
+  return storage.get();
 }
```
Comment (darkweak) [inline frontend/src/service/instance.tsx]: IMHO this can be removed.

**Review by darkweak — CHANGES_REQUESTED** [review]

File: frontend/src/service/instance.tsx [hunk 15]
```diff
@@ -1,13 +1,16 @@
...
-const instanceStorage = new InstanceData();
-
-export function setInstances(instances: ReadonlyArray<InstanceType>) {
-  instanceStorage.set(instances);
+export function setInstances(
+  storage: IStorage,
+  instances: ReadonlyArray<InstanceType>,
+) {
+  storage.set(instances);
 }
```
Comment (darkweak) [inline frontend/src/service/instance.tsx]: IMHO this can be removed

**Review by darkweak — CHANGES_REQUESTED** [review]

File: frontend/src/context/StorageProvider.tsx [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+  const { storage } = useContext(storageContext);
+
+  return storage;
+};
+
+export const useSelectStorage = () => {
+  const { storage, setStorage } = useContext(storageContext);
+
+  return { storage, setStorage };
+};
```
Comment (darkweak) [inline frontend/src/context/StorageProvider.tsx]: Because you're returning all props, you can just return your useContext.

**Review by darkweak — CHANGES_REQUESTED** [review]

**Review by darkweak — APPROVED** [review]

**Pull request merged**